dvc/stage/run.py

```diff
@@ -145,48 +170,109 @@ def run_stage(
+    proc: subprocess.Popen
+    done: threading.Event
+    killed: threading.Event
+    signal_path: str
+    AWAIT: ClassVar[float] = 1.0
+
+
+class Monitor(ABC):
+    def __init__(self, config: MonitorConfig):
+        self.config = config
+        self.monitor_thread = threading.Thread(
+            target=_monitor_loop, args=(self.config,),
+        )
+
+    def __enter__(self):
+        logger.debug(
+            "Monitoring stage '%s' with cmd process '%d'",
+            self.config.stage,
+            self.config.proc.pid,
+        )
+        self.monitor_thread.start()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.config.done.set()
+        self.monitor_thread.join()
+
+    @property
+    @abstractmethod
+    def SIGNAL_FILE(self) -> str:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def error_cls(self):
+        raise NotImplementedError
+
+    def signal_path(self, stage):
+        return os.path.join(stage.repo.tmp_dir, self.SIGNAL_FILE)
+
+
+class CheckpointMonitor(Monitor):
+    SIGNAL_FILE = "DVC_CHECKPOINT"
+    error_cls = CheckpointKilledError
+
+    def __init__(self, stage, callback_func, proc):
+        super().__init__(
+            MonitorConfig(
+                "checkpoint",
+                stage,
+                functools.partial(_run_callback, stage, callback_func),
+                proc,
+                threading.Event(),
+                threading.Event(),
+                self.signal_path(stage),
+            )
+        )
+
+
+class LiveMonitor(Monitor):
+    SIGNAL_FILE = "DVC_LIVE"
+    error_cls = LiveKilledError
+
+    def __init__(self, out, proc):
+        super().__init__(
+            MonitorConfig(
+                "live",
+                out.stage,
+                functools.partial(create_summary, out),
+                proc,
+                threading.Event(),
+                threading.Event(),
+                self.signal_path(out.stage),
+            )
+        )
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # make sure all data is visualied after training
+        self.config.task()
+        super().__exit__(exc_type, exc_val, exc_tb)
+
+
+def _monitor_loop(config: MonitorConfig):
     while True:
-        if os.path.exists(signal_path):
+        if os.path.exists(config.signal_path):
             try:
-                _run_callback(stage, callback_func)
+                config.task()
             except Exception:  # pylint: disable=broad-except
                 logger.exception(
-                    "Error generating checkpoint, %s will be aborted", stage
+                    "Error running '%s' task, '%s' will be aborted",
+                    config.name,
+                    config.stage,
                 )
-                _kill(proc)
-                killed.set()
+                _kill(config.proc)
+                config.killed.set()
             finally:
                 logger.debug("Remove checkpoint signal file")
```

Comment:
This message should probably be generalized now

Reply:
Also feels like the top level `_...` methods here should go into the base `Monitor` as class methods now (w/the checkpoint callback one in `CheckpointMonitor`)